### PR TITLE
DEPR: Deprecate week, weekofyear in Series.dt,DatetimeIndex

### LINKED ISSUE #33503: DEPR: weekofyear
Now that `isocalendar` is available on `Timestamp`, `DatetimeIndex` and the `dt` accessor, `weekofyear` already exists in `isocalendar` and can be deprecated.

### PR DESCRIPTION
**Pull request by mgmarino**
Closes https://github.com/pandas-dev/pandas/issues/33503

This PR is a followup to #33220 and implements what was discussed there.

A few comments:

- `DatetimeIndex.isocalendar` does not set the index on the returned dataframe,
  leading to some cumbersome syntax.  This may be worth changing, please let me
  know what you think.
- Should we also deprecate `Timestamp.week/weekofyear`?

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/groupby/transform/test_transform.py
Comment (mgmarino) [inline pandas/tests/groupby/transform/test_transform.py:1030]: This is the cumbersome syntax I referenced in the PR message.

File: pandas/core/indexes/accessors.py
Comment (jreback) [inline pandas/core/indexes/accessors.py:263]: you can put this import at the top of the file (same everywhere)

File: pandas/core/indexes/accessors.py
Comment (jreback) [inline pandas/core/indexes/accessors.py:273]: we do it like this now?

File: pandas/tests/arrays/test_datetimes.py
Comment (jreback) [inline pandas/tests/arrays/test_datetimes.py:431]: at the issue number as a comment

File: pandas/tests/indexes/datetimes/test_misc.py
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_misc.py:379]: same

File: pandas/tests/series/test_datetime_values.py
Comment (jreback) [inline pandas/tests/series/test_datetime_values.py:692]: same

**Review by jreback — CHANGES_REQUESTED** [review]
this looks fine to me

**Comment by jreback** [comment]
> DatetimeIndex.isocalendar does not set the index on the returned dataframe,
leading to some cumbersome syntax. This may be worth changing, please let me
know what you think.

hmm you mean the index of the original dataframe? I *think* we do this on the accessors, so we *should* do this.

File: pandas/tests/groupby/transform/test_transform.py
Comment (jbrockmendel) [reply pandas/tests/groupby/transform/test_transform.py:1030]: `pd.to_datetime(stocks.index)` is just `dates`, ditto `stocks.index`, so this can become `dates.isocalendar().set_index(dates)["week"]`

**Comment by jbrockmendel** [comment]
> DatetimeIndex.isocalendar does not set the index on the returned dataframe,
leading to some cumbersome syntax

agreed

File: pandas/core/indexes/datetimes.py
Comment (mgmarino) [reply pandas/core/indexes/datetimes.py:813]: Do you mean the deprecation should be handled by the underlying DatetimeArray?  I started trying to do it that way.  The reason for doing it this way is to get the stacklevel right for the warning.  If we are ok with having it "broken" for a direct call to `DatetimeArray.week`, then I can go that route.

File: pandas/core/indexes/accessors.py
Comment (mgmarino) [reply pandas/core/indexes/accessors.py:272]: Yes, in the current `week` accessor the array is converted to a float and masked if there are nans.    This reproduces that behavior instead of passing back a `UInt32` series.

**Comment by mgmarino** [comment]
> hmm you mean the index of the original dataframe? I think we do this on the accessors, so we should do this.

Yes, I mean that the DataFrame returned by DatetimeIndex.isocalendar() should probably have its index set to the DatetimeIndex.  This is done e.g. in Series.dt.isocalendar(): the returned DataFrame has it's index set to that of the Series.

For Index classes, the [current behavior](https://github.com/pandas-dev/pandas/blob/3a5ae505bcec7541a282114a89533c01a49044c0/pandas/core/indexes/extension.py#L60) of `inherit_names` when wrapping is to *convert* the result of the delegated call to the same *type* if possible (i.e. DatetimeIndex => DatetimeIndex) and to fall back to a generic Index if not.


> > DatetimeIndex.isocalendar does not set the index on the returned dataframe,
> > leading to some cumbersome syntax
> > 
> 
> agreed

Ok, the question would then be where to do this:

- It can be done in `inherit_names`, but the code [there](https://github.com/pandas-dev/pandas/blob/3a5ae505bcec7541a282114a89533c01a49044c0/pandas/core/indexes/extension.py#L60) would need to be adapted to handle a DataFrame (or something more generic?) and this is currently the only case for that.  
- Or the delegation can be done "manually" in an explicitly defined`DatetimeIndex.isocalendar` method.

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
I am personally not sure a deprecation is needed, see https://github.com/pandas-dev/pandas/issues/33503

File: pandas/core/indexes/datetimes.py
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py:813]: it will almost always be seen by users via the DatetimeIndex, so i think thats the case to prioritize getting the stacklevel right for

**Comment by mgmarino** [comment]
Build failing due to numba version.  Will wait for #33686 or patch from @mroeschke

File: pandas/core/indexes/datetimes.py
Comment (mgmarino) [reply pandas/core/indexes/datetimes.py]: Ok, I moved the deprecation In `DatetimeIndex` to only be in `DatetimeArray` and optimized the stacklevel appropriately.

**Comment by mgmarino** [comment]
@jbrockmendel @jreback 
Sorry for the delay, I believe all open comments have been addressed.  Tests green.

Regarding the setting of the index, I would like to do that in a separate PR if there are no objections.

File: pandas/tests/scalar/test_nat.py
Comment (jreback) [inline pandas/tests/scalar/test_nat.py:69]: we should properly parameterize these tests (if you would like to make an issue would be great), PR even better :->

**Review by jreback — APPROVED** [review]

File: pandas/core/arrays/datetimes.py
Comment (jbrockmendel) [inline pandas/core/arrays/datetimes.py:1322]: this should return ndarray here and be wrapped on DatetimeIndex to return Int64Index

Comment (mgmarino) [reply pandas/core/arrays/datetimes.py]: Of course, thanks.

File: pandas/tests/scalar/test_nat.py
Comment (mgmarino) [reply pandas/tests/scalar/test_nat.py:69]: Im happy to make another PR, will wait until this is merged.

**Review by mroeschke — APPROVED** [review]

**Comment by mgmarino** [comment]
@mroeschke @jreback @jbrockmendel Just checking in here to see if anything else needs to be done re discussion/PR?  Thanks!

**Comment by mgmarino** [comment]
@mroeschke @jreback I would like to ping one more time, to try to keep this from getting too stale.  If more discussion needs to be add about this, I'm totally fine with that, but I would like to also address the two issues that came up in the PR here.  If this needs more time to merge, then I would go ahead and address those already.

**Comment by mroeschke** [comment]
Sorry for the delay @mgmarino. I think everything looked good to me a while back. Looks like there's a merge conflict, but once that's resolved I think this should be good

**Review by mroeschke — APPROVED** [review]
LGTM @jreback

**Comment by jreback** [comment]
yeah this is ok, ping on green. @jorisvandenbossche unless any objections here.

**Comment by mgmarino** [comment]
Build failure due to a version bump of flake in the CI (now 3.8.1, last succeeded with 3.7.9).

**Comment by mgmarino** [comment]
@jreback back to green

**Comment by jreback** [comment]
this lgtm. we can merge on green.

**Comment by mroeschke** [comment]
Thanks for sticking with this @mgmarino. Great PR as always!

**Pull request merged**
